Update collaboration-engine documentation to 3.1 for Vaadin 14.7 (v14) (#616)
vaadin/docs: articles/ce/collaboration-message-list.asciidoc

controls for submitting new messages.
Here's an example of setting them up:

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/ce/MessageListDocumentation.java[tags=message-list-and-input,indent=0]
----
.Example classes
NOTE: `User` and `UserService` are application-specific example classes,
and not part of the Collaboration Engine API.

The `CollaborationMessageList` constructor takes two arguments: the information about
the end user associated with this session, and a topic id.
The user info is used to render the user name and avatar in the messages submitted by this user.
Refer to the <<collaboration-avatar-group#ce.configuring-avatars, `CollaborationAvatarGroup` documentation>>
for more details on how the avatars are rendered, or how to load images from a backend with the `setImageProvider` method.

The topic id works in the same way as for other Collaboration Engine features.
The data in Collaboration Engine is shared among those users who are connected to the same topic.
With these components, you can consider the topic as a chat room.
In the example above, the topic id is hard-coded as "general", which means that every

> user in this view sees each others' messages.

Reviewer: ```suggestion
user in this view sees the other users' messages.
```

for example when the user makes a selection in the UI.

`CollaborationMessageInput` takes the message list component as its constructor argument.
It hooks up with the list, to submit messages to its current topic.
When the list component has `null` topic id, no messages are displayed, and the connected
message input will be automatically disabled.

After constructing the components, the last required step is to place them somewhere in your view.
Here's an example of placing the message list and input on top of each other in a layout with a defined size:

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/ce/MessageListDocumentation.java[tags=message-list-layout,indent=0]
----

== Persisting Messages

The topic data in Collaboration Engine is stored only in the application memory.
This applies also to messages in a `CollaborationMessageList`, which means that
all the submitted messages will be lost when restarting the server.

To solve this issue, you can configure `CollaborationMessageList` to store the messages
in your own backend.
This is done by implementing the `CollaborationMessagePersister` interface.
It specifies two operations:

> - Fetch all messages in a given topic, which have been submitted after a given timestamp.

Reviewer: ```suggestion
- Fetch all messages in a given topic which have been submitted after a given timestamp.
```

- Store a new message.

In the following example implementation,
you can see how the persister is mapping between the `CollaborationMessage` objects
and the application's own `Message` entities, and integrating with the service layer:

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/ce/MyMessagePersister.java[tags=message-list-persister,indent=0]
----

.Example classes
NOTE: `MessageService`, `UserService`, `Message` and `User` are application-specific example classes,
and not part of the Collaboration Engine API.

.Initialize a persister from lambdas
TIP: You can also create a message persister instance with lambdas by using the static
`CollaborationMessagePersister.fromCallbacks()` method.

It's best to configure your database to insert the creation/insertion timestamp of the entity.
This ensures that the messages are ordered by their timestamps in the database.
If you use the timestamp provided by `CollaborationMessage::getTime` (based on the JVM clock),

> you take a risk of skipping some inserted messages in rare edge cases.

Reviewer: ```suggestion
you risk skipping some inserted messages in rare edge cases.
```

For example, if you use Hibernate to create the database table,
you can include an automatic insertion timestamp in your entity class
with the following annotation:
[source,java]
----
@Column(columnDefinition="TIMESTAMP DEFAULT CURRENT_TIMESTAMP")
private Instant time;
----

> You can set the message persister in the message list's constructor:

Reviewer: ```suggestion
You can set the message persister in the constructor of the message list:
```

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/ce/MessageListDocumentation.java[tags=message-list-ctor-persister,indent=0]
----

When a persister is provided, `CollaborationMessageList` takes care of keeping
the component state in sync with your backend.
When the first message list connects to a topic after starting the application,
it fetches all the messages from the persister

> (making a `FetchQuery` with Unix epoch as the `getSince()` timestamp).

Reviewer: ```suggestion
(doing a `FetchQuery` with Unix epoch as the `getSince()` timestamp).
```
